server/app/scanpy_engine/scanpy_engine.py

```diff
@@ -214,51 +225,53 @@ def layout(self, df):
+        only the top N vars will be returned.
+        :param df1: from filter_cells, dataframe containing first set of observations
+        :param df2: from filter_cells, dataframe containing second set of observations
+        :param topN: Limit results to top N (Top var mode only)
+        :return: top genes, stats and expression values for variables
         """
-        cells_idx_1 = np.in1d(self.data.obs["cell_name"], cell_list_1)
-        cells_idx_2 = np.in1d(self.data.obs["cell_name"], cell_list_2)
-        expression_1 = self.data.X[cells_idx_1, :]
-        expression_2 = self.data.X[cells_idx_2, :]
-        diff_exp = stats.ttest_ind(expression_1, expression_2)
-        # TODO break this up into functions
-        set1 = np.logical_and(diff_exp.pvalue < pval, diff_exp.statistic > 0)
-        set2 = np.logical_and(diff_exp.pvalue < pval, diff_exp.statistic < 0)
-        stat1 = diff_exp.statistic[set1]
-        stat2 = diff_exp.statistic[set2]
-        sort_set1 = np.argsort(stat1)[::-1]
-        sort_set2 = np.argsort(stat2)
-        pval1 = diff_exp.pvalue[set1][sort_set1]
-        pval2 = diff_exp.pvalue[set2][sort_set2]
-        mean_ex1_set1 = np.mean(expression_1[:, set1], axis=0)[sort_set1]
-        mean_ex2_set1 = np.mean(expression_2[:, set1], axis=0)[sort_set1]
-        mean_ex1_set2 = np.mean(expression_1[:, set2], axis=0)[sort_set2]
-        mean_ex2_set2 = np.mean(expression_2[:, set2], axis=0)[sort_set2]
-        mean_diff1 = mean_ex1_set1 - mean_ex2_set1
-        mean_diff2 = mean_ex1_set2 - mean_ex2_set2
-        genes_cellset_1 = self.data.var_names[set1][sort_set1]
-        genes_cellset_2 = self.data.var_names[set2][sort_set2]
-        return {
-            "celllist1": {
-                "topgenes": genes_cellset_1.tolist()[:num_genes],
-                "mean_expression_cellset1": mean_ex1_set1.tolist()[:num_genes],
-                "mean_expression_cellset2": mean_ex2_set1.tolist()[:num_genes],
-                "pval": pval1.tolist()[:num_genes],
-                "ave_diff": mean_diff1.tolist()[:num_genes]
-            },
-            "celllist2": {
-                "topgenes": genes_cellset_2.tolist()[:num_genes],
-                "mean_expression_cellset1": mean_ex1_set2.tolist()[:num_genes],
-                "mean_expression_cellset2": mean_ex2_set2.tolist()[:num_genes],
-                "pval": pval2.tolist()[:num_genes],
-                "ave_diff": mean_diff2.tolist()[:num_genes]
-            },
-        }
+        # If not the same genes, test is wrong!
+        if np.any(df1.var.index != df2.var.index):
+            raise ValueError("Variables ares not the same in set1 and set2")
+
+        # If not all genes, they used a var filter
+        if df1.var.shape[0] < self.gene_count:
+            mode = DiffExpMode("var filter")
+            if top_n:
+                raise Warning("Top N was specified but will not be used in 'Var Filter' mode")
```

Comment:
Do you want to raise a warning?

Possibly you want `warnings.warn("Top N was specified but will not be used in 'Var Filter' mode")`